File: client/php/revani.php
<?php

class RevaniClient {
    private $host;
    private $port;
    private $secure;
    private $_socket;
    private $_sessionKey;
    private $_accountID;
    private $_projectName;
    private $_projectID;

    public $account;
    public $project;
    public $data;
    public $storage;
    public $livekit;
    public $pubsub;

    public function __construct(string $host, int $port = 16897, bool $secure = true) {
        $this->host = $host;
        $this->port = $port;
        $this->secure = $secure;
        $this->account = new RevaniAccount($this);
        $this->project = new RevaniProject($this);
        $this->data = new RevaniData($this);
        $this->storage = new RevaniStorage($this);
        $this->livekit = new RevaniLivekit($this);
        $this->pubsub = new RevaniPubSub($this);
    }

    public function connect() {
        $protocol = $this->secure ? "ssl" : "tcp";
        $remote = $protocol . "://" . $this->host . ":" . $this->port;
        $context = stream_context_create([
            "ssl" => [
                "verify_peer" => false,
                "verify_peer_name" => false,
            ]
        ]);

        $this->_socket = stream_socket_client($remote, $errno, $errstr, 30, STREAM_CLIENT_CONNECT, $context);
        
        if (!$this->_socket) {
            throw new Exception("Connection failed: $errstr");
        }
    }

    public function execute(array $command, bool $useEncryption = true): array {
        $payload = ($useEncryption && $this->_sessionKey !== null)
            ? ['encrypted' => $this->_encrypt(json_encode($command))]
            : $command;

        $jsonPayload = json_encode($payload);
        $length = strlen($jsonPayload);
        $header = pack('N', $length);

        fwrite($this->_socket, $header);
        fwrite($this->_socket, $jsonPayload);

        $responseHeader = fread($this->_socket, 4);
        if (!$responseHeader) {
            $this->disconnect();
            throw new Exception("Connection closed");
        }

        $unpackHeader = unpack('Nlength', $responseHeader);
        $responseLength = $unpackHeader['length'];
        
        $responsePayload = "";
        while (strlen($responsePayload) < $responseLength) {
            $chunk = fread($this->_socket, $responseLength - strlen($responsePayload));
            if (!$chunk) break;
            $responsePayload .= $chunk;
        }

        $json = json_decode($responsePayload, true);

        if (isset($json['encrypted']) && $this->_sessionKey !== null) {
            $decrypted = $this->_decrypt($json['encrypted']);
            return json_decode($decrypted, true);
        }

        return $json;
    }

    private function _encrypt(string $text): string {
        $wrapper = json_encode([
            "payload" => $text,
            "ts" => (int)(microtime(true) * 1000)
        ]);

        $salt = random_bytes(16);
        $saltBase64 = base64_encode($salt);
        
        $keyBytes = hash('sha256', $this->_sessionKey . $saltBase64, true);
        $iv = random_bytes(16);
        
        $ciphertext = openssl_encrypt($wrapper, 'aes-256-gcm', $keyBytes, OPENSSL_RAW_DATA, $iv, $tag);

        return $saltBase64 . ":" . base64_encode($iv) . ":" . base64_encode($ciphertext . $tag);
    }

    private function _decrypt(string $encryptedData): string {
        $parts = explode(':', $encryptedData);
        $saltBase64 = $parts[0];
        $iv = base64_decode($parts[1]);
        $cipherAndTag = base64_decode($parts[2]);
        
        $tag = substr($cipherAndTag, -16);
        $ciphertext = substr($cipherAndTag, 0, -16);

        $keyBytes = hash('sha256', $this->_sessionKey . $saltBase64, true);
        
        $decrypted = openssl_decrypt($ciphertext, 'aes-256-gcm', $keyBytes, OPENSSL_RAW_DATA, $iv, $tag);
        
        $wrapper = json_decode($decrypted, true);
        return $wrapper['payload'];
    }

    public function setSession(string $key) { $this->_sessionKey = $key; }
    public function setAccount(string $id) { $this->_accountID = $id; }
    public function setProject(string $name, ?string $id) {
        $this->_projectName = $name;
        $this->_projectID = $id;
    }

    public function getAccountID(): string { return $this->_accountID ?? ""; }
    public function getProjectName(): string { return $this->_projectName ?? ""; }
    public function getProjectID(): string { return $this->_projectID ?? ""; }

    public function disconnect() {
        if ($this->_socket) {
            fclose($this->_socket);
        }
        $this->_socket = null;
        $this->_sessionKey = null;
        $this->_accountID = null;
        $this->_projectName = null;
        $this->_projectID = null;
    }
}

class RevaniAccount {
    private $_client;
    public function __construct($client) { $this->_client = $client; }

    public function create(string $email, string $password, ?array $extraData = null): array {
        return $this->_client->execute([
            'cmd' => 'account/create',
            'email' => $email,
            'password' => $password,
            'data' => $extraData ?? new stdClass(),
        ], false);
    }

    public function login(string $email, string $password): bool {
        $res = $this->_client->execute([
            'cmd' => 'auth/login',
            'email' => $email,
            'password' => $password,
        ], false);

        if (isset($res['status']) && $res['status'] == 200) {
            $this->_client->setSession($res['session_key']);
            $idRes = $this->_client->execute([
                'cmd' => 'account/get-id',
                'email' => $email,
                'password' => $password,
            ], false);

            if (isset($idRes['status']) && $idRes['status'] == 200) {
                $this->_client->setAccount($idRes['data']['id']);
            }
            return true;
        }
        return false;
    }

    public function getData(): array {
        return $this->_client->execute([
            'cmd' => 'account/get-data',
            'id' => $this->_client->getAccountID(),
        ]);
    }
}

class RevaniProject {
    private $_client;
    public function __construct($client) { $this->_client = $client; }

    public function use(string $projectName): array {
        $res = $this->_client->execute([
            'cmd' => 'project/exist',
            'accountID' => $this->_client->getAccountID(),
            'projectName' => $projectName,
        ]);

        if (isset($res['status']) && $res['status'] == 200) {
            $this->_client->setProject($projectName, $res['id']);
        }
        return $res;
    }

    public function create(string $projectName): array {
        $res = $this->_client->execute([
            'cmd' => 'project/create',
            'accountID' => $this->_client->getAccountID(),
            'projectName' => $projectName,
        ]);
        if (isset($res['status']) && $res['status'] == 200) {
            $this->_client->setProject($projectName, $res['data']['id']);
        }
        return $res;
    }
}

class RevaniData {
    private $_client;
    public function __construct($client) { $this->_client = $client; }

    public function add(string $bucket, string $tag, array $value): array {
        return $this->_client->execute([
            'cmd' => 'data/add',
            'accountID' => $this->_client->getAccountID(),
            'projectName' => $this->_client->getProjectName(),
            'bucket' => $bucket,
            'tag' => $tag,
            'value' => $value,
        ]);
    }

    public function get(string $bucket, string $tag): array {
        return $this->_client->execute([
            'cmd' => 'data/get',
            'projectID' => $this->_client->getProjectID(),
            'bucket' => $bucket,
            'tag' => $tag,
        ]);
    }

    public function query(string $bucket, array $query): array {
        return $this->_client->execute([
            'cmd' => 'data/query',
            'accountID' => $this->_client->getAccountID(),
            'projectName' => $this->_client->getProjectName(),
            'bucket' => $bucket,
            'query' => $query,
        ]);
    }

    public function update(string $bucket, string $tag, $newValue): array {
        return $this->_client->execute([
            'cmd' => 'data/update',
            'projectID' => $this->_client->getProjectID(),
            'bucket' => $bucket,
            'tag' => $tag,
            'newValue' => $newValue,
        ]);
    }

    public function delete(string $bucket, string $tag): array {
        return $this->_client->execute([
            'cmd' => 'data/delete',
            'projectID' => $this->_client->getProjectID(),
            'bucket' => $bucket,
            'tag' => $tag,
        ]);
    }
}

class RevaniStorage {
    private $_client;
    public function __construct($client) { $this->_client = $client; }

    public function upload(string $fileName, array $bytes, bool $compress = false): array {
        return $this->_client->execute([
            'cmd' => 'storage/upload',
            'accountID' => $this->_client->getAccountID(),
            'projectName' => $this->_client->getProjectName(),
            'fileName' => $fileName,
            'bytes' => $bytes,
            'compress' => $compress,
        ]);
    }

    public function download(string $fileId): array {
        return $this->_client->execute([
            'cmd' => 'storage/download',
            'accountID' => $this->_client->getAccountID(),
            'projectName' => $this->_client->getProjectName(),
            'fileId' => $fileId,
        ]);
    }

    public function delete(string $fileId): array {
        return $this->_client->execute([
            'cmd' => 'storage/delete',
            'accountID' => $this->_client->getAccountID(),
            'projectName' => $this->_client->getProjectName(),
            'fileId' => $fileId,
        ]);
    }
}

class RevaniLivekit {
    private $_client;
    public function __construct($client) { $this->_client = $client; }

    public function init(string $host, string $apiKey, string $apiSecret): array {
        return $this->_client->execute([
            'cmd' => 'livekit/init',
            'host' => $host,
            'apiKey' => $apiKey,
            'apiSecret' => $apiSecret,
        ]);
    }

    public function autoConnect(): array {
        return $this->_client->execute([
            'cmd' => 'livekit/connect',
            'accountID' => $this->_client->getAccountID(),
            'projectName' => $this->_client->getProjectName(),
        ]);
    }

    public function createToken(string $roomName, string $userID, string $userName, bool $isAdmin = false): array {
        return $this->_client->execute([
            'cmd' => 'livekit/create-token',
            'roomName' => $roomName,
            'userID' => $userID,
            'userName' => $userName,
            'isAdmin' => $isAdmin,
        ]);
    }

    public function createRoom(string $roomName, int $timeout = 10, int $maxUsers = 50): array {
        return $this->_client->execute([
            'cmd' => 'livekit/create-room',
            'roomName' => $roomName,
            'emptyTimeoutMinute' => $timeout,
            'maxUsers' => $maxUsers,
        ]);
    }
}

class RevaniPubSub {
    private $_client;
    public function __construct($client) { $this->_client = $client; }

    public function subscribe(string $topic, string $clientId): array {
        return $this->_client->execute([
            'cmd' => 'pubsub/subscribe',
            'accountID' => $this->_client->getAccountID(),
            'projectName' => $this->_client->getProjectName(),
            'clientId' => $clientId,
            'topic' => $topic,
        ]);
    }

    public function publish(string $topic, array $data, ?string $clientId = null): array {
        return $this->_client->execute([
            'cmd' => 'pubsub/publish',
            'accountID' => $this->_client->getAccountID(),
            'projectName' => $this->_client->getProjectName(),
            'topic' => $topic,
            'data' => $data,
            'clientId' => $clientId,
        ]);
    }

    public function unsubscribe(string $topic, string $clientId): array {
        return $this->_client->execute([
            'cmd' => 'pubsub/unsubscribe',
            'clientId' => $clientId,
            'topic' => $topic,
        ]);
    }
}

File: client/python/revani.py
# 
# Copyright (C) 2026 JeaFriday (https://github.com/JeaFrid/Revani)
# 
# Licensed under the GNU Affero General Public License v3.0 (AGPL-3.0).
# Project: Revani
#
import asyncio
import json
import struct
import base64
import time
import hashlib
import os
import ssl
from cryptography.hazmat.primitives.ciphers.aead import AESGCM

class RevaniClient:
    def __init__(self, host, port=16897, secure=True):
        self.host = host
        self.port = port
        self.secure = secure
        self._reader = None
        self._writer = None
        self._session_key = None
        self._account_id = None
        self._project_name = None
        self._project_id = None
        self.account = RevaniAccount(self)
        self.project = RevaniProject(self)
        self.data = RevaniData(self)
        self.storage = RevaniStorage(self)
        self.livekit = RevaniLivekit(self)
        self.pubsub = RevaniPubSub(self)

    async def connect(self):
        if self.secure:
            context = ssl.create_default_context()
            context.check_hostname = False
            context.verify_mode = ssl.CERT_NONE
            self._reader, self._writer = await asyncio.open_connection(
                self.host, self.port, ssl=context
            )
        else:
            self._reader, self._writer = await asyncio.open_connection(
                self.host, self.port
            )

    async def execute(self, command, use_encryption=True):
        if use_encryption and self._session_key:
            payload = {'encrypted': self._encrypt(json.dumps(command))}
        else:
            payload = command

        bytes_data = json.dumps(payload).encode('utf-8')
        header = struct.pack('>I', len(bytes_data))
        
        self._writer.write(header)
        self._writer.write(bytes_data)
        await self._writer.drain()

        header_resp = await self._reader.readexactly(4)
        length = struct.unpack('>I', header_resp)[0]
        payload_resp = await self._reader.readexactly(length)
        
        response_json = json.loads(payload_resp.decode('utf-8'))

        if isinstance(response_json, dict) and 'encrypted' in response_json and self._session_key:
            decrypted = self._decrypt(response_json['encrypted'])
            return json.loads(decrypted)
        
        return response_json

    def _encrypt(self, text):
        wrapper = json.dumps({
            "payload": text,
            "ts": int(time.time() * 1000)
        })
        
        salt_bytes = os.urandom(16)
        salt_b64 = base64.b64encode(salt_bytes).decode('utf-8')
        
        key_material = (self._session_key + salt_b64).encode('utf-8')
        key_hash = hashlib.sha256(key_material).digest()
        
        iv = os.urandom(16)
        aesgcm = AESGCM(key_hash)
        ciphertext = aesgcm.encrypt(iv, wrapper.encode('utf-8'), None)
        
        iv_b64 = base64.b64encode(iv).decode('utf-8')
        cipher_b64 = base64.b64encode(ciphertext).decode('utf-8')
        
        return f"{salt_b64}:{iv_b64}:{cipher_b64}"

    def _decrypt(self, encrypted_data):
        parts = encrypted_data.split(':')
        salt_b64 = parts[0]
        iv = base64.b64decode(parts[1])
        ciphertext = base64.b64decode(parts[2])
        
        key_material = (self._session_key + salt_b64).encode('utf-8')
        key_hash = hashlib.sha256(key_material).digest()
        
        aesgcm = AESGCM(key_hash)
        decrypted_bytes = aesgcm.decrypt(iv, ciphertext, None)
        
        wrapper = json.loads(decrypted_bytes.decode('utf-8'))
        return wrapper['payload']

    def set_session(self, key):
        self._session_key = key

    def set_account(self, account_id):
        self._account_id = account_id

    def set_project(self, name, project_id):
        self._project_name = name
        self._project_id = project_id

    @property
    def account_id(self):
        return self._account_id or ""

    @property
    def project_name(self):
        return self._project_name or ""

    @property
    def project_id(self):
        return self._project_id or ""

    async def disconnect(self):
        if self._writer:
            self._writer.close()
            await self._writer.wait_closed()
        self._reader = None
        self._writer = None
        self._session_key = None
        self._account_id = None
        self._project_name = None
        self._project_id = None

class RevaniAccount:
    def __init__(self, client):
        self._client = client

    async def create(self, email, password, extra_data=None):
        return await self._client.execute({
            'cmd': 'account/create',
            'email': email,
            'password': password,
            'data': extra_data or {},
        }, use_encryption=False)

    async def login(self, email, password):
        res = await self._client.execute({
            'cmd': 'auth/login',
            'email': email,
            'password': password,
        }, use_encryption=False)

        if res.get('status') == 200:
            self._client.set_session(res['session_key'])
            id_res = await self._client.execute({
                'cmd': 'account/get-id',
                'email': email,
                'password': password,
            }, use_encryption=False)

            if id_res.get('status') == 200:
                self._client.set_account(id_res['data']['id'])
            return True
        return False

    async def get_data(self):
        return await self._client.execute({
            'cmd': 'account/get-data',
            'id': self._client.account_id,
        })

class RevaniProject:
    def __init__(self, client):
        self._client = client

    async def use(self, project_name):
        res = await self._client.execute({
            'cmd': 'project/exist',
            'accountID': self._client.account_id,
            'projectName': project_name,
        })

        if res.get('status') == 200:
            self._client.set_project(project_name, res.get('id'))
        return res

    async def create(self, project_name):
        res = await self._client.execute({
            'cmd': 'project/create',
            'accountID': self._client.account_id,
            'projectName': project_name,
        })
        if res.get('status') == 200:
            self._client.set_project(project_name, res['data']['id'])
        return res

class RevaniData:
    def __init__(self, client):
        self._client = client

    async def add(self, bucket, tag, value):
        return await self._client.execute({
            'cmd': 'data/add',
            'accountID': self._client.account_id,
            'projectName': self._client.project_name,
            'bucket': bucket,
            'tag': tag,
            'value': value,
        })

    async def get(self, bucket, tag):
        return await self._client.execute({
            'cmd': 'data/get',
            'projectID': self._client.project_id,
            'bucket': bucket,
            'tag': tag,
        })

    async def query(self, bucket, query_map):
        return await self._client.execute({
            'cmd': 'data/query',
            'accountID': self._client.account_id,
            'projectName': self._client.project_name,
            'bucket': bucket,
            'query': query_map,
        })

    async def update(self, bucket, tag, new_value):
        return await self._client.execute({
            'cmd': 'data/update',
            'projectID': self._client.project_id,
            'bucket': bucket,
            'tag': tag,
            'newValue': new_value,
        })

    async def delete(self, bucket, tag):
        return await self._client.execute({
            'cmd': 'data/delete',
            'projectID': self._client.project_id,
            'bucket': bucket,
            'tag': tag,
        })

class RevaniStorage:
    def __init__(self, client):
        self._client = client

    async def upload(self, file_name, byte_list, compress=False):
        return await self._client.execute({
            'cmd': 'storage/upload',
            'accountID': self._client.account_id,
            'projectName': self._client.project_name,
            'fileName': file_name,
            'bytes': byte_list,
            'compress': compress,
        })

    async def download(self, file_id):
        return await self._client.execute({
            'cmd': 'storage/download',
            'accountID': self._client.account_id,
            'projectName': self._client.project_name,
            'fileId': file_id,
        })

    async def delete(self, file_id):
        return await self._client.execute({
            'cmd': 'storage/delete',
            'accountID': self._client.account_id,
            'projectName': self._client.project_name,
            'fileId': file_id,
        })

class RevaniLivekit:
    def __init__(self, client):
        self._client = client

    async def init(self, host, api_key, api_secret):
        return await self._client.execute({
            'cmd': 'livekit/init',
            'host': host,
            'apiKey': api_key,
            'apiSecret': api_secret,
        })

    async def auto_connect(self):
        return await self._client.execute({
            'cmd': 'livekit/connect',
            'accountID': self._client.account_id,
            'projectName': self._client.project_name,
        })

    async def create_token(self, room_name, user_id, user_name, is_admin=False):
        return await self._client.execute({
            'cmd': 'livekit/create-token',
            'roomName': room_name,
            'userID': user_id,
            'userName': user_name,
            'isAdmin': is_admin,
        })

    async def create_room(self, room_name, timeout=10, max_users=50):
        return await self._client.execute({
            'cmd': 'livekit/create-room',
            'roomName': room_name,
            'emptyTimeoutMinute': timeout,
            'maxUsers': max_users,
        })

class RevaniPubSub:
    def __init__(self, client):
        self._client = client

    async def subscribe(self, topic, client_id):
        return await self._client.execute({
            'cmd': 'pubsub/subscribe',
            'accountID': self._client.account_id,
            'projectName': self._client.project_name,
            'clientId': client_id,
            'topic': topic,
        })

    async def publish(self, topic, data, client_id=None):
        return await self._client.execute({
            'cmd': 'pubsub/publish',
            'accountID': self._client.account_id,
            'projectName': self._client.project_name,
            'topic': topic,
            'data': data,
            'clientId': client_id,
        })

    async def unsubscribe(self, topic, client_id):
        return await self._client.execute({
            'cmd': 'pubsub/unsubscribe',
            'clientId': client_id,
            'topic': topic,
        })